### PR TITLE
Honour output filename when SIDE_MODULE is used

### PR DESCRIPTION
**Pull request by sbc100**
Previously the wasm output file would always be forced to `.wasm`
even if some other extensions was requested.

Importantly this meant that `emcc -s SIDE_MODULE -o foo.so` would
not output `foo.so` as expected.

Split out from #11799

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -505,3 +505,4 @@ a license to everyone to use it as detailed in LICENSE.)
 * Paul Peny <pmpp.pub@gmail.com>
 * Prashanth Nethi <prashant@adobe.com>
 * Max Weisel <max@maxweisel.com>
+* Georg Rottensteiner <georg@georg-rottensteiner.de>
```
Comment (kripken) [inline AUTHORS:508]: this may need to be rebased, as this and other changes look irrelevant?

Comment (sbc100) [reply AUTHORS]: Done.

File: emcc.py [hunk 2]
```diff
@@ -1112,7 +1112,13 @@ def add_link_flag(i, f):
       shared.Settings.EXPORT_ES6 = 1
       shared.Settings.MODULARIZE = 1
 
-    wasm_target = unsuffixed(target) + '.wasm' # might not be used
+    if shared.Settings.SIDE_MODULE or final_suffix in WASM_ENDINGS:
+      # If the user asks directly for a wasm file then this *is* the target
```
Comment (sbc100) [inline emcc.py:1116]: As you mentioned in a previous incarnation of this PR `final_suffix in WASM_ENDINGS` is currently redundant here, but I like it for clarity and it means its safe against adding new WASM_ENDINGS in the future.

File: emcc.py [hunk 3]
```diff
@@ -1881,7 +1887,7 @@ def get_clang_command_asm(src_file):
         return 0
 
       # Precompiled headers support
-      if has_header_inputs:
+      if has_header_inputs or 'header' in language_mode:
```
Comment (kripken) [inline emcc.py:1890]: is this change unrelated? i can't connect it.

File: emcc.py [hunk 4]
```diff
@@ -2225,12 +2227,12 @@ def get_final():
...
-      if final_suffix in JS_ENDINGS:
-        js_target = target
-      elif final_suffix in WASM_ENDINGS:
+      if final_suffix in WASM_ENDINGS:
         js_target = misc_temp_files.get(suffix='.js').name
-      else:
+      elif final_suffix == '.html':
         js_target = unsuffixed(target) + '.js'
+      else:
+        js_target = target
```
Comment (kripken) [inline emcc.py:2235]: how is SIDE_MODULE handled by this if-else chain? It seems like the final suffix would be `.so` if  the user asked for that, so we'd reach the final `else`, and then `js_target = foo.so`? I must be misunderstanding how this works, sorry...

File: emcc.py (hunk 3 above)
Comment (sbc100) [reply emcc.py:1890]: This technically a more correct behaviour so we definitely want to land it.  I'm trying to remember why it is needed for this change .. it cause a test to fail without it.  I can check to confirm which one now.

File: emcc.py (hunk 2 above)
Comment (sbc100) [reply emcc.py:1116]: We don't this far with SIDE_MODULE (see return statement just above) since there is no js generated for that case.

File: emcc.py (hunk 4 above)
Comment (sbc100) [reply emcc.py:2235]: We don't this far with SIDE_MODULE (see return statement just above) since there is no js generated for that case.

File: emcc.py (hunk 3 above)
Comment (sbc100) [reply emcc.py]: Tried reverting this.  Will land it later as part of a separate change (if this passes)

File: emcc.py (hunk 4 above)
Comment (kripken) [reply emcc.py:2235]: Ah, that makes sense then. But then this is unrelated to the PR, isn't it? (if the PR is for SIDE_MODULE, and this code is not relevant for SIDE_MODULE - am I missing something?)

**Comment by sbc100** [comment]
OK this change should be good to go now.

File: emcc.py (hunk 4 above)
Comment (kripken) [reply emcc.py]: Not sure about the temp directory specifically, but standalone mode still has a JS file, which can be used to load and run the wasm on the web (while the wasm can still be used standalone off the web).

File: emcc.py [hunk 5]
```diff
@@ -2008,12 +2014,8 @@ def compile_source_file(i, input_file):
 
     if link_to_object:
       with ToolchainProfiler.profile_block('linking to object file'):
-        # We have a specified target (-o <target>), which is not JavaScript or HTML, and
-        # we have multiple files: Link them
-        if shared.Settings.SIDE_MODULE:
-          exit_with_error('SIDE_MODULE must only be used when compiling to an executable shared library, and not when emitting an object file.  That is, you should be emitting a .wasm file (for wasm) or a .js file (for asm.js). Note that when compiling to a typical native suffix for a shared library (.so, .dylib, .dll; which many build systems do) then Emscripten emits an object file, which you should then compile to .wasm or .js with SIDE_MODULE.')
```
Comment (kripken) [inline emcc.py:2014]: maybe worth keeping an assert for not reaching here with SIDE_MODULE?

**Review by kripken — APPROVED** [review]

File: emcc.py (hunk 5 above)
Comment (sbc100) [reply emcc.py:2014]: I actually think is probably OK to read here with SIDE_MODULE set now.

For example I think both of these are not completely unreasonable:

`emcc -r out.o in.o -s SIDE_MODULE`
`emcc -shared libout.so in.o -s SIDE_MODULE`

Comment (sbc100) [reply emcc.py:2014]: The `-r` and the `-shared` basically mean full linking is not done, but that doesn't need to mean SIDE_MODULE can't be specified.  So this change relaxed this a little.
